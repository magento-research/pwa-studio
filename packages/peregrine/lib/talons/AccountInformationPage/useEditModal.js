import { useAppContext } from '../../context/app';

export const useEditModal = () => {
    const [{ drawer }, { closeDrawer }] = useAppContext();
    const isOpen = drawer === 'accountInformation.edit';

    return {
        handleClose: closeDrawer,
        isOpen
    };
};
