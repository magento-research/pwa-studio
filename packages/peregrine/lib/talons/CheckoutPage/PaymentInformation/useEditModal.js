import { useState, useCallback } from 'react';

/**
 * Talon to handle checkout page's payment information edit modal.
 *
 * @param {Function} props.onClose callback to be called when the modal's close or cancel button is clicked.
 *
 * @returns {
 *   isLoading: Boolean,
 *   shouldRequestPaymentNonce: Boolean,
 *   handleClose: Function,
 *   handleUpdate: Function,
 *   handlePaymentSuccess: Function,
 *   handleDropinReady: Function,
 *   handlePaymentError: Function
 * }
 */
export const useEditModal = props => {
    const { onClose } = props;

    /**
     * Definitions
     */

    const [isLoading, setIsLoading] = useState(true);
    const [shouldRequestPaymentNonce, setShouldRequestPaymentNonce] = useState(
        false
    );

    /**
     * Helper Functions
     */

    const handleClose = useCallback(() => {
        onClose();
    }, [onClose]);

    const handleUpdate = useCallback(() => {
        setShouldRequestPaymentNonce(true);
    }, [setShouldRequestPaymentNonce]);

    const handlePaymentSuccess = useCallback(() => {
        onClose();
    }, [onClose]);

    const handlePaymentError = useCallback(() => {
        setShouldRequestPaymentNonce(false);
    }, []);

    const handleDropinReady = useCallback(() => {
        setIsLoading(false);
    }, [setIsLoading]);

    return {
        isLoading,
        shouldRequestPaymentNonce,
        handleClose,
        handleUpdate,
        handlePaymentSuccess,
        handlePaymentError,
        handleDropinReady
    };
};
