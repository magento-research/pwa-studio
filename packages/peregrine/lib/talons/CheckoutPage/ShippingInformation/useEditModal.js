import { useAppContext } from '../../../context/app';

export const useEditModal = () => {
    const [{ drawer }, { closeDrawer }] = useAppContext();
    const isOpen = drawer === 'shipping.edit';

    return {
        handleClose: closeDrawer,
        isOpen
    };
};
