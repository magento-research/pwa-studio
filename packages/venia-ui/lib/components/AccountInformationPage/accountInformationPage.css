.root {
    display: grid;
    padding: 2.5rem 3rem;
    row-gap: 2rem;
}

.title {
    justify-self: center;
    font-family: var(--venia-global-fontFamily-serif);
    font-weight: var(--venia-global-fontWeight-bold);
}

.lineItems {
    display: grid;
    grid-gap: 0.75rem;
    grid-template-columns: 1fr auto;
    line-height: 1.5rem;
}

.lineItemLabel {
    font-weight: var(--venia-global-fontWeight-bold);
}

.editInformationButton {
    composes: root_normalPriority from '../Button/button.css';
    margin: auto;
}

@media (min-width: 961px) {
    .cardArea {
        display: grid;
        gap: 1rem;
        grid-template-columns: 1fr 1fr 1fr;
    }
}

@media (max-width: 960px) {
    .lineItems {
        grid-template-columns: 1fr;
    }

    .lineItemButton {
        text-align: center;
    }
}
