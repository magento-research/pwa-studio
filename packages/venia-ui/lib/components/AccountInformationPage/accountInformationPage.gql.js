import gql from 'graphql-tag';

/* export const SET_NEWSLETTER_SUBSCRIPTION = gql`
    mutation SetNewsletterSubscription($isSubscribed: Boolean!) {
        updateCustomer(input: { is_subscribed: $isSubscribed }) {
            customer {
                id
                is_subscribed
            }
        }
    }
`; */

export const GET_CUSTOMER_INFORMATION = gql`
    query GetCustomerInformation {
        customer {
            id
            firstname
            lastname
            email
            # password
        }
    }
`;

export default {
    /* mutations: {
        setNewsletterSubscriptionMutation: SET_NEWSLETTER_SUBSCRIPTION
    }, */
    queries: {
        getCustomerInformationQuery: GET_CUSTOMER_INFORMATION
    }
};
