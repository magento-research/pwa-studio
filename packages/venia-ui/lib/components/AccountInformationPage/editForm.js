import React, { Fragment } from 'react';
import { func, shape, string, bool } from 'prop-types';

import { mergeClasses } from '../../classify';
import Field from '../Field';
import TextInput from '../TextInput';
import LinkButton from '../LinkButton';
import Password from '../Password';
import combine from '../../util/combineValidators';
import {
    hasLengthAtLeast,
    isRequired,
    validatePassword,
    validateDifferentCurrentPassword
} from '../../util/formValidators';
import defaultClasses from './editForm.css';

const EditForm = props => {
    const {
        classes: propClasses,
        handleActivePassword,
        isChangingPassword
    } = props;

    const classes = mergeClasses(defaultClasses, propClasses);

    const passwordArea = !isChangingPassword ? (
        <Fragment>
            <div className={classes.passwordLabel}>
                <Field id="password" label="Password">
                    <span>{'**********'}</span>
                </Field>
            </div>
            <div className={classes.changePasswordButtonContainer}>
                <LinkButton
                    classes={classes.changePasswordButton}
                    type="button"
                    onClick={handleActivePassword}
                >
                    {'Change Password'}
                </LinkButton>
            </div>
        </Fragment>
    ) : (
        <Fragment>
            <div className={classes.password}>
                <Password
                    fieldName="password"
                    label="Current Password"
                    validate={isRequired}
                    autoComplete="current-password"
                    isToggleButtonHidden={false}
                />
            </div>
            <div className={classes.newPassword}>
                <Password
                    fieldName="newPassword"
                    label="New Password"
                    validate={combine([
                        isRequired,
                        [hasLengthAtLeast, 8],
                        validatePassword,
                        validateDifferentCurrentPassword
                    ])}
                    isToggleButtonHidden={false}
                />
            </div>
        </Fragment>
    );

    return (
        <div className={classes.root}>
            <div className={classes.firstname}>
                <Field id="firstname" label="First Name">
                    <TextInput field="firstname" validate={isRequired} />
                </Field>
            </div>
            <div className={classes.lastname}>
                <Field id="lastname" label="Last Name">
                    <TextInput field="lastname" validate={isRequired} />
                </Field>
            </div>
            <div className={classes.email}>
                <Field id="email" label="Email">
                    <TextInput field="email" validate={isRequired} />
                </Field>
            </div>
            {passwordArea}
        </div>
    );
};

export default EditForm;

EditForm.propTypes = {
    classes: shape({
        root: string,
        field: string,
        email: string,
        firstname: string,
        lastname: string,
        buttons: string,
        passwordLabel: string,
        changePasswordButtonContainer: string,
        changePasswordButton: string,
        password: string,
        newPassword: string
    }),
    handleActivePassword: func,
    isChangingPassword: bool
};
