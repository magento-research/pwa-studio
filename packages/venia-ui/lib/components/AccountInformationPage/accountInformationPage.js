import React, { Fragment } from 'react';
import { Redirect } from '@magento/venia-drivers';
import { useAccountInformationPage } from '@magento/peregrine/lib/talons/AccountInformationPage/useAccountInformationPage';

import { mergeClasses } from '../../classify';
import Button from '../Button';
import { Message } from '../Field';
import { Title } from '../Head';
import { fullPageLoadingIndicator } from '../LoadingIndicator';
import EditModal from './editModal';
import defaultClasses from './accountInformationPage.css';
import AccountInformationPageOperations from './accountInformationPage.gql.js';

const AccountInformationPage = props => {
    const classes = mergeClasses(defaultClasses, props.classes);

    const talonProps = useAccountInformationPage({
        ...AccountInformationPageOperations
    });

    const {
        loadDataError,
        formErrors,
        handleSubmit,
        initialValues,
        isDisabled,
        isSignedIn,
        handleEditInformation,
        activeChangePassword,
        handleChangePassword
    } = talonProps;

    if (!isSignedIn) {
        return <Redirect to="/" />;
    }

    const errorMessage = loadDataError ? (
        <Message>
            {
                'Something went wrong. Please refresh and try again.'
            }
        </Message>
    ) : null;

    let pageContent = null;
    if (!initialValues) {
        return fullPageLoadingIndicator;
    } else {
        const { customer } = initialValues;

        pageContent = (
            <Fragment>
                <div className={classes.cardArea}>
                    <div className={classes.lineItems}>
                        <span className={classes.lineItemLabel}>{'Name'}</span>
                        <span className={classes.lineItemValue}>{`${
                            customer.firstname
                        } ${customer.lastname}`}</span>
                        <span className={classes.lineItemLabel}>{'Email'}</span>
                        <span className={classes.lineItemValue}>
                            {customer.email}
                        </span>
                        <span className={classes.lineItemLabel}>
                            {'Password'}
                        </span>
                        <span className={classes.lineItemValue}>
                            {'***********'}
                        </span>
                        <span className={classes.lineItemLabel} />
                        <span className={classes.lineItemValue}>
                            <Button
                                className={classes.editInformationButton}
                                disabled={false}
                                onClick={handleEditInformation}
                                priority="normal"
                            >
                                {'Edit'}
                            </Button>
                        </span>
                    </div>
                </div>
                <EditModal
                    informationData={customer}
                    activeChangePassword={activeChangePassword}
                    handleChangePassword={handleChangePassword}
                    handleSubmit={handleSubmit}
                    isDisabled={isDisabled}
                    formErrors={formErrors}
                />
            </Fragment>
        );
    }

    return (
        <div className={classes.root}>
            <Title>{`Account Information - ${STORE_NAME}`}</Title>
            <h1 className={classes.title}>{'Account Information'}</h1>
            {errorMessage ? errorMessage : pageContent}
        </div>
    );
};

export default AccountInformationPage;
