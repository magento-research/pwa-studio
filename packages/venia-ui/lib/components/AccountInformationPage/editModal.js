import React, { useCallback } from 'react';
import { object, shape, string, bool, array, func } from 'prop-types';
import { X as CloseIcon } from 'react-feather';
import { useEditModal } from '@magento/peregrine/lib/talons/AccountInformationPage/useEditModal';

import { mergeClasses } from '../../classify';
import Icon from '../Icon';
import { Portal } from '../Portal';
import EditForm from './editForm';
import defaultClasses from './editModal.css';

const EditModal = props => {
    const {
        classes: propClasses,
        informationData,
        isDisabled,
        formErrors,
        activeChangePassword,
        handleChangePassword,
        handleSubmit
    } = props;
    const talonProps = useEditModal();
    const { handleClose, isOpen } = talonProps;

    const classes = mergeClasses(defaultClasses, propClasses);
    const rootClass = isOpen ? classes.root_open : classes.root;

    const onCancelModal = useCallback(() => {
        handleClose();
        handleChangePassword(false);
    }, [handleClose, handleChangePassword]);

    // Unmount the form to force a reset back to original values on close
    const bodyElement = isOpen ? (
        <EditForm
            informationData={informationData}
            handleSubmit={handleSubmit}
            activeChangePassword={activeChangePassword}
            isDisabled={isDisabled}
            handleCancel={handleClose}
            handleChangePassword={handleChangePassword}
            formErrors={formErrors}
            onCancelModal={onCancelModal}
        />
    ) : null;

    return (
        <Portal>
            <aside className={rootClass}>
                <div className={classes.header}>
                    <span className={classes.headerText}>
                        {'Edit Account Information'}
                    </span>
                    <button
                        className={classes.closeButton}
                        onClick={handleClose}
                    >
                        <Icon src={CloseIcon} />
                    </button>
                </div>
                <div className={classes.body}>{bodyElement}</div>
            </aside>
        </Portal>
    );
};

export default EditModal;

EditModal.propTypes = {
    classes: shape({
        root: string,
        root_open: string,
        closeButton: string,
        body: string,
        header: string,
        headerText: string
    }),
    informationData: object,
    isDisabled: bool,
    formErrors: array,
    handleSubmit: func,
    handleChangePassword: func,
    activeChangePassword: bool
};
