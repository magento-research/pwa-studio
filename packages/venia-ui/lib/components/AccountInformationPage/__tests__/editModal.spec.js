import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useEditModal } from '@magento/peregrine/lib/talons/AccountInformationPage/useEditModal';

import EditModal from '../editModal';

jest.mock('@magento/peregrine/lib/talons/AccountInformationPage/useEditModal');
jest.mock('../../../classify');
jest.mock('../../Portal', () => ({
    Portal: props => <portal-mock>{props.children}</portal-mock>
}));
jest.mock('../editForm', () => 'EditForm');

const handleClose = jest.fn().mockName('handleClose');
const handleChangePassword = jest.fn().mockName('handleChangePassword');
const handleSubmit = jest.fn().mockName('handleSubmit');

const props = {
    classes: {},
    informationData: {
        firstname: 'Huy',
        lastname: 'Kon',
        email: 'huykon@gmail.com'
    },
    isDisabled: false,
    formErrors: [],
    activeChangePassword: false,
    handleChangePassword,
    handleSubmit
};

test('renders open modal', () => {
    useEditModal.mockReturnValueOnce({
        handleClose,
        isOpen: true
    });

    const tree = createTestInstance(<EditModal {...props} />);
    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders closed modal', () => {
    useEditModal.mockReturnValueOnce({
        handleClose,
        isOpen: false
    });

    const tree = createTestInstance(<EditModal {...props} />);
    expect(tree.toJSON()).toMatchSnapshot();
});
