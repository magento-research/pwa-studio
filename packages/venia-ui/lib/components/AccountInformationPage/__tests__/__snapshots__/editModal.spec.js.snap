// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders closed modal 1`] = `
<portal-mock>
  <aside
    className="root"
  >
    <div
      className="header"
    >
      <span
        className="headerText"
      >
        Edit Account Information
      </span>
      <button
        className="closeButton"
        onClick={[MockFunction handleClose]}
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      className="body"
    />
  </aside>
</portal-mock>
`;

exports[`renders open modal 1`] = `
<portal-mock>
  <aside
    className="root_open"
  >
    <div
      className="header"
    >
      <span
        className="headerText"
      >
        Edit Account Information
      </span>
      <button
        className="closeButton"
        onClick={[MockFunction handleClose]}
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      className="body"
    >
      <EditForm
        activeChangePassword={false}
        formErrors={Array []}
        handleCancel={[MockFunction handleClose]}
        handleChangePassword={[MockFunction handleChangePassword]}
        handleSubmit={[MockFunction handleSubmit]}
        informationData={
          Object {
            "email": "huykon@gmail.com",
            "firstname": "Huy",
            "lastname": "Kon",
          }
        }
        isDisabled={false}
        onCancelModal={[Function]}
      />
    </div>
  </aside>
</portal-mock>
`;
